### PR TITLE
perf: optimize hashing

### PR DESCRIPTION
**Pull request by ananas-block**
Changes:
- improved performance of hashing for both token and compressed account data by omitting values that are not used
  - for safety added prefixes to values that are less than 32 bytes
  - the hashing order MUST NOT be changed without very careful review because optional 32 bytes values can also be confused for each other and are not prefixed the way it currently is it is ok because 31 byte values are nested between values that are prefixed
  - added methods that allow for hashed inputs (before we were rehashing for example mint every time
- introduced a buffer vector for hashed account to pass them between the input and output compressed account hashing functions to avoid rehashing of Pubkeys

### CODE REVIEW

File: programs/compressed-pda/src/invoke/verify_state_proof.rs [hunk 1]
```diff
@@ -103,14 +111,64 @@ pub fn hash_input_compressed_accounts<
...
+                        &inputs.output_compressed_accounts[j].owner.to_bytes(),
+                    )
+                    .unwrap()
+                    .0;
+                    hashed_pubkeys.push((inputs.output_compressed_accounts[j].owner, hashed_owner));
+                    hashed_owner
+                }
+            };
+        }
+        msg!("owner: {:?}", owner_pubkey);
```
Comment (vadorovsky) [inline programs/compressed-pda/src/invoke/verify_state_proof.rs:166]: Is it necessary, or it was just a debug log?

File: programs/compressed-token/src/token_data.rs [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+#[repr(u8)]
+pub enum AccountState {
+    Initialized,
+    Frozen,
+}
+
+#[derive(Debug, PartialEq, Eq, AnchorSerialize, AnchorDeserialize, Clone, Copy)]
+pub struct TokenData {
+    /// The mint associated with this account
+    pub mint: Pubkey,
```
Comment (vadorovsky) [inline programs/compressed-token/src/token_data.rs:19]: `TokenData` and `TokenData` are almost identical, I can spot only one different field. Any chance we could move the common fields to a new struct and include it as fields?

Comment (ananas-block) [reply programs/compressed-token/src/token_data.rs:19]: yeah this is on purpose. Anchor ts enums didn't work out of the box, hence the workaround to have a TokenDataClient struct that is explicit about the enum being a u8.
We should figure out how to handle the enum on ts side and remove the TokenDataClient then.

File: programs/compressed-pda/src/invoke/verify_state_proof.rs (hunk 1 above)
Comment (ananas-block) [reply programs/compressed-pda/src/invoke/verify_state_proof.rs]: removed

**Review by vadorovsky — APPROVED** [review]

**Pull request merged**